projects/lights_control/test/test_lights_blinker.c

```diff
@@ -0,0 +1,161 @@
+
+  Event e;
+  // Wait for the blinker_1 initial event to fire. This happens almost instantly as the event gets
+  // raised upon call to lights_blinker_on.
+  while (event_process(&e) != STATUS_CODE_OK) {
+  }
+  TEST_ASSERT_EQUAL(e.id, LIGHTS_BLINKER_TEST_EVENT_1);
+  TEST_ASSERT_EQUAL(e.data, LIGHTS_BLINKER_STATE_ON);
+
+  TEST_ASSERT_OK(lights_blinker_on(&blinker_2, DURATION_LONG, LIGHTS_BLINKER_TEST_EVENT_2));
+  TEST_ASSERT_EQUAL(blinker_2.duration_ms, DURATION_LONG);
+  TEST_ASSERT_EQUAL(LIGHTS_BLINKER_STATE_ON, blinker_2.state);
+  TEST_ASSERT_NOT_EQUAL(blinker_2.timer_id, SOFT_TIMER_INVALID_TIMER);
+
+  // Wait for the blinker_2 initial event to fire.
+  while (event_process(&e) != STATUS_CODE_OK) {
+  }
+  TEST_ASSERT_EQUAL(e.id, LIGHTS_BLINKER_TEST_EVENT_2);
+  TEST_ASSERT_EQUAL(e.data, LIGHTS_BLINKER_STATE_ON);
+
+  // Now we make sure blinker 1 timer goes off before blinker 2
+  while (event_process(&e) != STATUS_CODE_OK) {
+  }
+  TEST_ASSERT_EQUAL(e.id, LIGHTS_BLINKER_TEST_EVENT_1);
+  TEST_ASSERT_EQUAL(e.data, LIGHTS_BLINKER_STATE_OFF);
+
+  // Wait for the second blinker
+  while (event_process(&e) != STATUS_CODE_OK) {
+  }
+  TEST_ASSERT_EQUAL(e.id, LIGHTS_BLINKER_TEST_EVENT_2);
+  TEST_ASSERT_EQUAL(e.data, LIGHTS_BLINKER_STATE_OFF);
+}
+
+void test_lights_blinker_on_uninitialized(void) {
+  // We shoudln't be able to call blinker_on on an uninitialized blinker, as uninitialized blinkers
+  // might have timer_id's that may be valid timer id's and may be in use by other modules in the
+  // application.
+  LightsBlinker blinker;
+  TEST_ASSERT_NOT_OK(lights_blinker_on(&blinker, DURATION_SHORT, LIGHTS_BLINKER_TEST_EVENT_1));
+}
+
+void test_lights_blinker_on_already_active(void) {
+  // lights_blinker_on requires that the blinker be inactive before calling.
+  LightsBlinker blinker;
+  TEST_ASSERT_OK(lights_blinker_init(&blinker));
+  TEST_ASSERT_OK(lights_blinker_on(&blinker, DURATION_SHORT, LIGHTS_BLINKER_TEST_EVENT_1));
+  TEST_ASSERT_NOT_OK(lights_blinker_on(&blinker, DURATION_SHORT, LIGHTS_BLINKER_TEST_EVENT_1));
+}
+
+void test_lights_blinker_off(void) {
+  // Making sure timer gets cancelled after we turn the blinker off
+  LightsBlinker blinker;
+  TEST_ASSERT_OK(lights_blinker_init(&blinker));
+  TEST_ASSERT_OK(lights_blinker_on(&blinker, DURATION_SHORT, LIGHTS_BLINKER_TEST_EVENT_1));
+  TEST_ASSERT_OK(lights_blinker_off(&blinker));
+  TEST_ASSERT_EQUAL(blinker.timer_id, SOFT_TIMER_INVALID_TIMER);
+  TEST_ASSERT_FALSE(soft_timer_inuse());
+}
+
+void test_lights_blinker_off_already_inactive(void) {
+  // lights_blinker_off requires that the blinker be an active blinker
+  LightsBlinker blinker;
+  TEST_ASSERT_OK(lights_blinker_init(&blinker));
+  TEST_ASSERT_OK(lights_blinker_on(&blinker, DURATION_SHORT, LIGHTS_BLINKER_TEST_EVENT_1));
+  TEST_ASSERT_OK(lights_blinker_off(&blinker));
+  TEST_ASSERT_NOT_OK(lights_blinker_off(&blinker));
+}
+
+void test_lights_blinker_reset(void) {
```

Comment:
What happens if you init and then reset?

Reply:
Ah good case! it shouldn't be possible, only "turned-on" blinkers can be reset.

Reply:
I'll add a test for this.

Reply:
Actually, `test_lights_blinker_reset_while_off` covers this, an initialized blinker is off by default.

Reply:
No, it doesn't. You turn it on and then off again.